exams/pearson.py

```diff
@@ -55,90 +67,147 @@ def _get_field_mapper(field):
     """
-    columns = [column for column, _ in fields]
-    field_mappers = [(column, _get_field_mapper(field)) for column, field in fields]
-    prefix_mapper = attrgetter(field_prefix) if field_prefix is not None else None
-
-    def _map_row(row):
-        if prefix_mapper:
-            row = prefix_mapper(row)
-        return {column: field_mapper(row) for column, field_mapper in field_mappers}
-
-    def _writer(file, rows):
-        tsv_writer = csv.DictWriter(
-            file,
-            columns,
+    def __init__(self, fields, field_prefix=None):
+        """
+        Initializes a new TSV writer
+
+        The first value of each fields tuple is the destination field name.
+        The second value is a str property path (e.g. "one.two.three") or
+        a callable that when passed a row returns a computed field value
+
+        Arguments:
+            fields (List): list of (str, str|callable) tuples
+            field_prefix (str): path prefix to prefix field lookups with
+        """
+        self.fields = OrderedDict(fields)
+        self.columns = self.fields.keys()
+        self.field_mappers = {column: _get_field_mapper(field) for column, field in self.fields.items()}
+        self.prefix_mapper = attrgetter(field_prefix) if field_prefix is not None else None
+
+    def _map_row(self, row):
+        """
+        Maps a row object to a row dict
+
+        Args:
+            row: the row to map to a dict
+
+        Returns:
+            dict:
+                row mapped to a dict using the field mappers
+        """
+        if self.prefix_mapper:
+            row = self.prefix_mapper(row)
+        return {column: field_mapper(row) for column, field_mapper in self.field_mappers.items()}
+
+    def write(self, tsv_file, rows):
+        """
+        Writes the rows to the designated file using the configured fields.
+
+        Invalid records are not written.
+
+        Arguments:
+            tsv_file: a file-like object to write the data to
+            rows: list of records to write to the tsv file
+
+        Returns:
+            (valid_record, invalid_records):
+                a tuple of which records were valid or invalid
+        """
+        file_writer = csv.DictWriter(
+            tsv_file,
+            self.columns,
             dialect=PEARSON_CSV_DIALECT,
             restval='',  # ensure we don't print 'None' into the file for optional fields
         )
 
-        tsv_writer.writeheader()
+        file_writer.writeheader()
 
         valid_rows, invalid_rows = [], []
 
         for row in rows:
             try:
-                tsv_writer.writerow(_map_row(row))
+                file_writer.writerow(self._map_row(row))
                 valid_rows.append(row)
             except InvalidTsvRow:
                 log.exception("Invalid tsv row")
                 invalid_rows.append(row)
 
         return (valid_rows, invalid_rows)
 
-    return _writer
 
+class CDDWriter(TSVWriter):
```

Comment:
@giocalitri Here is the revised class-based writers that I refactored based on your comment [here](https://github.com/mitodl/micromasters/pull/2221#discussion_r94607920).